rust/private/rustdoc_test.bzl

```diff
@@ -0,0 +1,147 @@
+# Copyright 2018 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("@io_bazel_rules_rust//rust:private/rustc.bzl", "CrateInfo", "DepInfo", "collect_deps", "get_lib_name")
+load("@io_bazel_rules_rust//rust:private/utils.bzl", "find_toolchain")
+
+def _rust_doc_test_impl(ctx):
+    if CrateInfo not in ctx.attr.dep:
+        fail("Expected rust library or binary.", "dep")
+
+    crate = ctx.attr.dep[CrateInfo]
+    rust_doc_test = ctx.outputs.executable
+
+    toolchain = find_toolchain(ctx)
+
+    dep_info = collect_deps(
+        [ctx.attr.dep],
+        toolchain,
+    )
+
+    # Construct rustdoc test command, which will be written to a shell script
+    # to be executed to run the test.
+    ctx.file_action(
+        output = rust_doc_test,
+        content = _build_rustdoc_test_script(toolchain, dep_info, crate),
+        executable = True,
+    )
+
+    # The test script compiles the crate and runs it, so it needs both compile and runtime inputs.
+    compile_inputs = (
+        crate.srcs +
+        [crate.output] +
+        dep_info.transitive_libs +
+        [toolchain.rust_doc] +
+        [toolchain.rustc] +
+        toolchain.rustc_lib +
+        toolchain.rust_lib +
+        toolchain.crosstool_files
+    )
+
+    runfiles = ctx.runfiles(
+        files = compile_inputs,
+        collect_data = True,
+    )
+    return struct(runfiles = runfiles)
+
+def dirname(path_str):
+    return "/".join(path_str.split("/")[:-1])
+
+def _build_rustdoc_test_script(toolchain, dep_info, crate):
+    """ Constructs the rustdoc script used to test `crate`. """
+
+    d = dep_info
+
+    # nb. Paths must be constructed wrt runfiles, so we construct relative link flags for doctest.
+    link_flags = []
+    link_search_flags = []
+
+    link_flags += ["--extern " + c.name + "=" + c.output.short_path for c in d.direct_crates]
```

Comment:
This was basically inlined when removing setup_deps, would say this bit is still fairly awkward.